# Propagating Errors

Often, we want to return errors from functions:

```rust
use std::io;
use std::io::Read;
use std::fs::File;

fn read_username_from_file() -> Result<String, io::Error> {
    let f = File::open("emoji.json");

    let mut f = match f {
        Ok(file) => file,
        Err(e) => return Err(e),
    };

    let mut s = String::new();

    match f.read_to_string(&mut s) {
        Ok(_) => Ok(s),
        Err(e) => Err(e),
    }
}
```

Above, we have to explicitly handle each error with a `match`.

Although this may feel similar to C, it's a practice that can become quite
verbose over time...
